src/modules/internal_combustion_engine_control/InternalCombustionEngineControl.cpp

```diff
@@ -0,0 +1,365 @@
+
+	internal_combustion_engine_status_s ice_status;
+	ice_status.state = static_cast<uint8_t>(_state);
+	ice_status.timestamp = now;
+	_internal_combustion_engine_status_pub.publish(ice_status);
+}
+
+bool InternalCombustionEngineControl::isEngineRunning(const hrt_abstime now)
+{
+	rpm_s rpm;
+	_rpm_sub.copy(&rpm);
+
+	const hrt_abstime rpm_timestamp = rpm.timestamp;
+
+	return (_param_ice_min_run_rpm.get() > FLT_EPSILON && (now < rpm_timestamp + 2_s)
+		&& rpm.rpm_estimate > _param_ice_min_run_rpm.get());
+}
+
+void InternalCombustionEngineControl::controlEngineRunning(float throttle_in)
+{
+	_ignition_on = true;
+	_choke_control = 0.f;
+	_starter_engine_control = 0.f;
+	_throttle_control = throttle_in;
+
+}
+
+void InternalCombustionEngineControl::controlEngineStop()
+{
+	_ignition_on = false;
+	_choke_control = _param_ice_stop_choke.get() ? 1.f : 0.f;
+	_starter_engine_control = 0.f;
+	_throttle_control = 0.f;
+}
+
+void InternalCombustionEngineControl::controlEngineFault()
+{
+	_ignition_on = false;
+	_choke_control = _param_ice_stop_choke.get() ? 1.f : 0.f;
+	_starter_engine_control = 0.f;
+	_throttle_control = 0.f;
+}
+
+void InternalCombustionEngineControl::controlEngineStartup(const hrt_abstime now)
+{
+	float ignition_delay = 0.f;
+	float choke_duration = 0.f;
+	const float starter_duration = _param_ice_starting_dur.get();
+
+	if (_starting_retry_cycle == 0) {
+		ignition_delay = math::max(_param_ice_ign_delay.get(), 0.f);
+
+		if (_param_ice_choke_st_dur.get() > FLT_EPSILON) {
+			choke_duration = _param_ice_choke_st_dur.get();
+		}
+	}
+
+	_ignition_on = true;
+	_throttle_control = _param_ice_strt_thr.get();
+	_choke_control = now < _state_start_time + (choke_duration + ignition_delay) * 1_s ? 1.f : 0.f;
+	_starter_engine_control = now > _state_start_time + (ignition_delay * 1_s) ? 1.f : 0.f;
+	const hrt_abstime cycle_timeout_duration = (ignition_delay + choke_duration + starter_duration) * 1_s;
+
+	if (now > _state_start_time + cycle_timeout_duration) {
+		// start resting timer if engine is not running
+		_starting_rest_time = now;
+		_starting_retry_cycle++;
+		PX4_INFO("ICE: Retry %i finished", _starting_retry_cycle);
+	}
+}
+
+bool InternalCombustionEngineControl::isStartingPermitted(const hrt_abstime now)
+{
+	return now > _starting_rest_time + DELAY_BEFORE_RESTARTING * 1_s;
+}
+
+bool InternalCombustionEngineControl::maximumRetriesReached()
+{
+	// First and only attempt
+	if (_param_ice_strt_retry.get() == 0) {
+		return _starting_retry_cycle > 0;
+	}
+
+	return _starting_retry_cycle >= _param_ice_strt_retry.get();
+}
+
+int InternalCombustionEngineControl::print_usage(const char *reason)
+{
+	if (reason) {
+		PX4_ERR("%s\n", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Description
+ICE controls.
+)DESCR_STR");
+
+	PRINT_MODULE_USAGE_NAME("internal_combustion_engine_control", "system");
+	PRINT_MODULE_USAGE_COMMAND_DESCR("start", "Start the background task");
```

Comment:
What background task?

Reply:
I think I copied that over. It doesn't add value so let's remove it.